Use CMake with a Makefile wrapper, instead of vanilla Make

modocache opened the pull request:
This makes good on #28 and #29.

I removed some of the `.PHONY` targets from your Makefile. I don't really know what those do, either... 😅  If you have any links handy, that'd be great!

TODO
- [ ] Get `make check` passing
- [ ] Decide which version of CMake to use/install for CI
- [ ] Decide which version of CMake to set as the project minimum
- [ ] Get CI builds working
- [ ] Review for parity with previous `Makefile`
- [x] Test that header changes cause rebuilds (#28)
- [ ] Test that tests can be built in parallel (#29)

modocache committed:
Use CMake

Define the project's build system using CMake. The Makefile is now
simply a convenient way of calling CMake.

All of the old commands still work, with the exception of the examples.

modocache committed:
[CMake] Add examples

Port the old examples/library_folders build commands from Make to CMake.

kastiglione commented on examples/library_folders/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+set(LIBRARY_FOLDERS_TARGET library_folders)
+
+file(GLOB library_folders_sources *.c)
+add_executable("${LIBRARY_FOLDERS_TARGET}" ${library_folders_sources})
```
To make `make` not automatically build this target, use [`EXCLUDE_FROM_ALL`](https://cmake.org/cmake/help/v3.0/command/add_executable.html).

kastiglione commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+project(Arbiter C CXX)
+
+cmake_minimum_required(VERSION 3.0)
```
Why not use a recent version, like 3.5 or 3.6.

Also, the docs say to put this first, even above the `project()` command.

modocache replied:
That's a good idea, although it looks like Travis CI doesn't have a very modern CMake installed by default on Linux machines.

jspahrsummers replied:
It's okay to install things from `apt`—most of the Travis toolchain is pretty old (e.g., C++ compilers only supporting C++11).

modocache replied:
(So, GitHub review comments don't go away when the code they're commenting on is gone? Weird. That's part of what I loved about vanilla GitHub comments.)

Done.

kastiglione commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+project(Arbiter C CXX)
```
`C` and `CXX` are default, is there value to being explicit?

modocache replied:
Oh cool, I guess not! Thanks!

modocache replied:
Done.

kastiglione commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.0)
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99 -pedantic -Wall -Wextra")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -pedantic -Wall -Wextra")
```
The `std` values can have default values set using `CMAKE_{C,CXX}_STANDARD`.

kastiglione replied:
I wonder why this form is idiomatic, and not `string(APPEND)`?

``` cmake
string(APPEND CMAKE_C_FLAGS " -pedantic -Wall -Wextra")
```

kastiglione replied:
Actually it turns out someone codemodded CMake's own repo to use `string(APPEND)` where applicable: https://github.com/Kitware/CMake/commit/5d0d980d9949daf596e10715d686adc95c1c232b

modocache replied:
Nice! Let's use that instead.

modocache replied:
Done.

modocache commented:
@kastiglione you have push access to my fork, feel free to make some of your suggested changes.

jspahrsummers commented:
I don't really know Make in detail either. 😉 

> Oh, also, I removed some of the `.PHONY` targets from your Makefile. I don't really know what those do…

Phony targets are those which should always be built, ignoring any dependencies. For example, `make check` should always invoke the test runner even if it was just run, because you probably intended to re-run the tests (not just rebuild some artifact).

I'll review the rest of this in more detail when I have time, just wanted to answer that question in case it's blocking. ☺️

jspahrsummers commented on Makefile:
```diff
@@ -1,60 +1,23 @@
-AR ?= ar
-RANLIB ?= ranlib
-XCODEBUILD ?= xcodebuild
+build := build
```
Convention is usually all-uppercase for Makefile variables.

modocache replied:
Done.

jspahrsummers commented on Makefile:
```diff
@@ -1,60 +1,23 @@
-LIBRARY = libArbiter.a
+all:
+	mkdir -p $(build)
+	$(cmake) -B$(build) -H. -G "Unix Makefiles" -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE
```
Does `cmake` have more verbose (spelled out) variants of these flags? If so, it'd be nice to use them for understandability.

modocache replied:
Totally agree. I'll find out and use detailed names if available.

modocache replied:
I couldn't find long versions of these options... 😞

jspahrsummers commented on Makefile:
```diff
@@ -1,60 +1,23 @@
-
-all: build
+check: all
+	set -e; for test in `find $(build) -name "*Test" -type f -perm +111`; do echo; echo "$$test"; ./$$test; done
```
Would you mind factoring this out into its own shell script, a la `script/cibuild`?

modocache replied:
Done.

jspahrsummers commented on test/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+endfunction()
+
+file(GLOB test_sources *.cpp)
+list(REMOVE_ITEM test_sources TestValue.cpp)
```
Hmm, I think I'd prefer an explicit list of tests rather than excluded files. It's common to have to list tests somewhere, and I don't want to disincentivize refactoring test code. ☺️

modocache replied:
That would be way easier to code, but I definitely assumed you'd want automatic test detection. What if a contributor renames a test and forgets to check that it ever runs?

Still, up to you. I'll change it to what I initially had, which involved listing the tests manually.

jspahrsummers replied:
Yeah, good point—I guess there are downsides either way. We'll keep an eye on it and can change it later if it causes problems.

kastiglione replied:
@jspahrsummers is your desire here still to have an explicitly managed list of tests?

jspahrsummers replied:
Mildly yes, but feel free to do what's easiest.

jspahrsummers commented on test/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+    PRIVATE "${ARBITER_TARGET}" gtest gtest_main)
+  target_include_directories("${test_name}"
+    PRIVATE
+      "${CMAKE_SOURCE_DIR}/external/googletest/googletest/include"
```
Is there a way to include this like `-isystem` does, to silence warnings in gtest headers like this?

```
In file included from /Users/jspahrsummers/GitHub/Arbiter/test/VersionTest.cpp:3:
/Users/jspahrsummers/GitHub/Arbiter/external/googletest/googletest/include/gtest/gtest.h:1392:11: warning: comparison of integers of different signs: 'const unsigned int' and 'const int'
      [-Wsign-compare]
  if (lhs == rhs) {
      ~~~ ^  ~~~
/Users/jspahrsummers/GitHub/Arbiter/external/googletest/googletest/include/gtest/gtest.h:1421:12: note: in instantiation of function template specialization
      'testing::internal::CmpHelperEQ<unsigned int, int>' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/Users/jspahrsummers/GitHub/Arbiter/test/VersionTest.cpp:12:3: note: in instantiation of function template specialization 'testing::internal::EqHelper<false>::Compare<unsigned int, int>' requested
      here
  EXPECT_EQ(version._major, 1);
  ^
/Users/jspahrsummers/GitHub/Arbiter/external/googletest/googletest/include/gtest/gtest.h:1924:63: note: expanded from macro 'EXPECT_EQ'
                      EqHelper<GTEST_IS_NULL_LITERAL_(val1)>::Compare, \
                                                              ^
1 warning generated.
```

modocache replied:
Done.

jspahrsummers commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+cmake_minimum_required(VERSION 3.0)
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99 -pedantic -Wall -Wextra")
```
Can you please re-add `-Wno-unused-parameter` here? It's for `library_folders`, because C doesn't permit unnamed parameters to designate lack of use.

modocache replied:
Oops, my bad. I had assumed I copied the flags over verbatim, but I guess I missed one. Will do!

modocache replied:
Done.

jspahrsummers reviewed (changes requested):
Thanks for doing this! 🙇

modocache committed:
[CMake] Use 3.6, add scripts

- Uses CMake 3.6, the most recent version.
- Adds scripts to install CMake 3.6 and run tests.
- Ports all C++ flags from the old Make build system.
- Include gtest headers as system headers.

modocache commented on Makefile:
```diff
@@ -1,60 +1,23 @@
-LIBRARY = libArbiter.a
+all:
+	mkdir -p $(build)
+	$(cmake) -B$(build) -H. -G "Unix Makefiles" -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE
```
Done.

kastiglione commented on script/ciinstall:
```diff
@@ -0,0 +1,12 @@
+
+set -e
+
+wget https://cmake.org/files/v3.6/cmake-3.6.0.tar.gz
```
Why 3.6.0 instead of latest 3.6.2.

modocache replied:
No reason, I didn't know 3.6.2 existed. I'll update.

kastiglione commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.6.2)
```
Should this stay at 3.6.0? Or is it good practice to keep this the same version as CI? When I commented that CI could use the latest version, I assumed without thinking that these could stay 3.6.0, since it seems unnecessary for force users to a specific patch release number.

modocache replied:
I don't have strong opinions here. 3.6.0 or 3.6.2, all good to me.

kastiglione replied:
What do you think is the right trade off:
1. Keep min version and CI version the same, and expect users to have this same version installed => which ensures no version related bugs
2. Set min version to 3.6.0 and ignore the patch number so that users can install any 3.6.x version for convenience => which risks hitting version related bugs

jspahrsummers replied:
I'd prefer option 1 until we hear from someone that it's a problem.

kastiglione committed:
Set EXCLUDE_FROM_ALL on library_folders target

kastiglione committed:
Add `examples` target back to Makefile

kastiglione commented on Makefile:
```diff
@@ -1,60 +1,28 @@
-TEST_INCLUDES = -isystem $(GTEST_DIR)/include -I$(GTEST_DIR) -Isrc/
+cmake:
+	mkdir -p $(BUILD)
+	$(CMAKE) -B$(BUILD) -H. -G "Unix Makefiles" -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE
```
I'd like to make this Makefile check for the presence of `ninja` and use that if available.

kastiglione committed:
Replace -std= flags with cmake properties

modocache commented on CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+
+set(CMAKE_C_STANDARD 99)
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_EXTENSIONS OFF) # -std=c++14 (otherwise cmake uses -std=gnu++14)
```
Nice.

kastiglione committed:
Use cmake's build tool mode

https://cmake.org/cmake/help/v3.6/manual/cmake.1.html#build-tool-mode

This allows the underlying build tool to change (e.g. using Ninja) but without
having to change invocations within the Makefile.

jspahrsummers commented:
Is this ready for another review? No rush—just not sure of its current status.

kastiglione commented:
I just added some TODOs to the description. This isn't quite done, but most of the todos shouldn't have much impact on the implementation as it is now.

Commit:
Merge branch 'master' into cmake

Commit:
Add src/ to include paths

To match upstream Makefile changes.

Commit:
Replace source globs with explicit lists

The docs for `file(GLOB...)` say:

    Note: We do not recommend using GLOB to collect a list of source files from
    your source tree. If no CMakeLists.txt file changes when a source is added
    or removed then the generated build system cannot know when to ask CMake to
    regenerate.

Commit:
Use concurrent builds with make

kastiglione commented:
@jspahrsummers, @modocache How do I find out which version of cmake is provided via APT packages? If it's not too old, I'm going to go with that for simplicity.

Besides getting CI setup, the remaining work is ensuring this does everything the current `Makefile` does, as well as maybe some small idiomatic improvements.

kastiglione commented:
so… `make check` has failures with this setup :grumpycat:

jspahrsummers commented:
I don't know how to check, exactly, except by looking at the Travis logs in detail. You can always add new `apt` sources too (like we do for GCC).

modocache commented:
I'm gonna close this: my heart's not in it anymore 😝 

@kastiglione if you still feel like landing this, please feel free to cherry-pick/amend any commits you'd like from `modocache:cmake`!

(Too bad GitHub doesn't allow people to "commandeer" pull requests, like Phabricator does.)
